build_runner/lib/src/daemon/daemon_builder.dart

```diff
@@ -132,12 +132,13 @@ class BuildRunnerDaemonBuilder implements DaemonBuilder {
     PackageGraph packageGraph,
     List<BuilderApplication> builders,
     SharedOptions sharedOptions,
+    BuildEnvironment overrideEnvironment,
   ) async {
     var expectedDeletes = Set<AssetId>();
     var outputStreamController = StreamController<ServerLog>();
 
     var environment = OverrideableEnvironment(
-        IOEnvironment(packageGraph,
+        overrideEnvironment ?? IOEnvironment(packageGraph,
```

Comment:
Not sure what the best API would be to combine these two would be.

Reply:
We could make an optional named arg, which is a factory that takes a BuildEnvironment and returns an OverrideableEnvironment: `{OverrideableEnvironment Function(BuildEnvironment) overrideEnvironment}`?

Then, you could conditionally wrap the environment in an additional layer:

``dart
if (overrideEnvironment != null) {
  environment = overrideEnvironment(original);
}
``